src/txn/txn_rollback_to_stable.c

```diff
@@ -110,20 +45,194 @@ __txn_abort_newer_update(
+    } else {
+        WT_ERR(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_NONE;
+        upd->durable_ts = WT_TS_NONE;
+        upd->start_ts = WT_TS_NONE;
+    }
+
+    /* If we don't yet have a modify structure, we'll need one. */
+    WT_ERR(__wt_page_modify_init(session, page));
+    mod = page->modify;
+
+    /* Allocate an update array as necessary. */
+    WT_PAGE_ALLOC_AND_SWAP(session, page, mod->mod_row_update, upd_entry, page->entries);
+
+    /* Set the WT_UPDATE array reference. */
+    upd_entry = &mod->mod_row_update[WT_ROW_SLOT(page, rip)];
+    upd->next = NULL;
+    WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, size, false));
+
+    /* Finally remove that update from history store. */
+    if (valid_update_found) {
+        __wt_hs_store_time_pair(session, WT_TS_NONE, WT_TXN_NONE);
+        WT_ERR(hs_cursor->remove(hs_cursor));
+        WT_STAT_CONN_INCR(session, txn_rollback_hs_removed);
+    }
+
+err:
+    __wt_scr_free(session, &key);
+    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &hs_value);
+    if (ret != 0)
+        __wt_free(session, upd);
+    WT_TRET(__wt_hs_cursor_close(session, &hs_cursor, session_flags));
+
+    return (ret);
+}
+
+/*
+ * __txn_abort_row_ondisk_kv --
+ *     Fix the on-disk row K/V version according to the given timestamp.
+ */
+static int
+__txn_abort_row_ondisk_kv(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip, wt_timestamp_t rollback_timestamp)
+{
+    WT_CELL_UNPACK *vpack, _vpack;
+    WT_DECL_RET;
+    WT_PAGE_MODIFY *mod;
+    WT_UPDATE *upd, **upd_entry;
+    size_t size;
+
+    vpack = &_vpack;
+    __wt_row_leaf_value_cell(session, page, rip, NULL, vpack);
+    if (vpack->start_ts > rollback_timestamp)
+        return __txn_abort_row_replace_with_hs_value(session, page, rip, rollback_timestamp);
+    else if (vpack->stop_ts != WT_TS_MAX && vpack->stop_ts > rollback_timestamp) {
+        /*
+         * Clear the remove operation from the key by inserting a TOMBSTONE with infinite
+         * visibility.
+         */
+        WT_RET(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_MAX;
+        upd->durable_ts = WT_TS_MAX;
+        upd->start_ts = WT_TS_MAX;
+    } else
+        /* Stable version according to the timestamp. */
+        return (0);
+
+    /* If we don't yet have a modify structure, we'll need one. */
+    WT_ERR(__wt_page_modify_init(session, page));
+    mod = page->modify;
+
+    /* Allocate an update array as necessary. */
+    WT_PAGE_ALLOC_AND_SWAP(session, page, mod->mod_row_update, upd_entry, page->entries);
+
+    /* Set the WT_UPDATE array reference. */
+    upd_entry = &mod->mod_row_update[WT_ROW_SLOT(page, rip)];
+    upd->next = NULL;
+    WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, size, false));
+    WT_STAT_CONN_INCR(session, txn_rollback_upd_aborted);
+    return (0);
+
+err:
+    __wt_free(session, upd);
```

Comment:
I'd recommend initializing `upd = NULL` at the start of the function and testing it here. If the code is ever modified, and a jump to `WT_ERR` added earlier in the function, it may be hard to spot the bug this would cause.

Reply:
Fixed.